File: Cpp/Cpp.DataStructures/src/structures/ChainedHashTable.cpp
#pragma once
#include "ChainedHashTable.h"
#include "math.h"
namespace DataStructures {
	void ChainedHashTable::Push(std::string input)
	{
		int id = GetId(GetHash(input));
		dataHolder[id].push_back(input);
		size++;
	}

	bool ChainedHashTable::Contain(std::string input)
	{
		if (IsEmpty()) {
			return false;
		}
		int id = GetId(GetHash(input));
		if (dataHolder[id].empty()) {
			return false;
		}
		for (const auto& elem : dataHolder[id]) {
			if (elem == input) {
				return true;
			}
		}
		return false;
	}

	size_t ChainedHashTable::Depth() const
	{
		if (IsEmpty()) {
			return 0;
		}
		size_t maxDepht = 0;
		for (auto vec : dataHolder) {
			maxDepht = std::max(vec.size(), maxDepht);
		}
		return size_t();
	}

	size_t ChainedHashTable::Count() const
	{
		return size;
	}

	bool ChainedHashTable::IsEmpty() const
	{
		return size == 0;
	}

	std::string ChainedHashTable::Print() const
	{
		std::ostringstream os;
		for (size_t i = 0; i < tableSize; i++)
		{
			os << '[' << i << "]:";
			if (dataHolder[i].empty()) {
				os << "[empty]"<<std::endl;
				continue;
			}
			for (const auto& elem : dataHolder[i]) {
				os << '[' << elem << ']';
			}
			os << std::endl;
		}
		return os.str();
	}

	unsigned int ChainedHashTable::GetHash(std::string input) const
	{
		unsigned int sum = 0;
		for (char ch : input) {
			sum += static_cast<int>(ch);
		}
		unsigned int hash = sum ^ input.size();
		return hash;
	}

	int ChainedHashTable::GetId(unsigned int hash) {
		return hash % tableSize;
	}
}

File: C#Code/Containers/ChainedHashTable.cs
using System.Text;

namespace Containers;

public class ChainedHashTable {
    private const int tableSize = 10;
    private int _currentSize = 0;
    private readonly List<string>[] _dataHolder;
    public bool IsEmpty => _currentSize == 0;
    public int Count => _currentSize;
    public ChainedHashTable() {
        _dataHolder = new List<string>[tableSize];
        for (int i = 0; i < _dataHolder.Length; i++) {
            _dataHolder[i] = new List<string>();
        }
    }
    public ChainedHashTable(params string[] words) : this() => InsertRange(words);
    public ChainedHashTable(IEnumerable<string> words) : this() => InsertRange(words);
    private uint GetHash(string value) =>
        value.Aggregate(0u, (hash, character) => hash += hash ^ character) ^ ((uint)value.Length * (uint)value.Length);
    private uint GetId(uint hash) =>
        hash % tableSize;
    public void Insert(string input) {
        _dataHolder[GetId(GetHash(input))].Add(input);
        _currentSize++;
    }
    public void InsertRange(IEnumerable<string> input) {
        foreach(var inputItem in input) {
            Insert(inputItem);
        }
    }
    public bool Contain(string input) => IsEmpty ? false : _dataHolder[GetId(GetHash(input))].Any(x => x == input);
    public void Remove(string input) {
        if (IsEmpty) {
            return;
        }
        _dataHolder[GetId(GetHash(input))].Remove(input);
        _currentSize--;
    }
    public void RemoveRange(IEnumerable<string> input) {
        foreach (var inputItem in input) {
            Remove(inputItem);
        }
    }
    public void Clear() {
        foreach (var list in _dataHolder) { 
            list.Clear();
        }
    }
    public override string ToString() {
        StringBuilder sb = new StringBuilder();
        sb.Append($"{base.ToString()}{{Count:{Count},TableSize:{tableSize},Table:\n");
        for (int index = 0; index < _dataHolder.Length; index++) {
            List<string> list = _dataHolder[index];
            sb.Append($"Chain[{index}]:");
            if (IsEmpty) {
                sb.Append("Empty");
            }
            list.ForEach(item => sb.Append($"[{item}]"));
            sb.Append('\n');
        }
        sb.Append('}');
        return sb.ToString();
    }
}